File: cpp/src/entropy/CMPredictor.cpp
/*
Copyright 2011-2017 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/


#include "../types.hpp"
#include "CMPredictor.hpp"

using namespace kanzi;

CMPredictor::CMPredictor()
{
    _ctx = 1;
    _run = 1;
    _idx = 8;
    _runMask = 0;
    _c1 = 0;
    _c2 = 0;

    for (int i = 0; i < 256; i++) {
        for (int j = 0; j <= 256; j++)
            _counter1[i][j] = 32768;

        for (int j = 0; j <= 16; j++) {
            _counter2[i + i][j] = j << 12;
            _counter2[i + i + 1][j] = j << 12;
        }

        _counter2[i + i][16] -= 16;
        _counter2[i + i + 1][16] -= 16;
    }

    _pc1 = _counter1[_ctx];
    _pc2 = _counter2[(_ctx << 1) | _runMask];
}

// Update the probability model
inline void CMPredictor::update(int bit)
{
    _ctx <<= 1;

    if (bit == 0) {
        _pc1[256] -= (_pc1[256] >> FAST_RATE);
        _pc1[_c1] -= (_pc1[_c1] >> MEDIUM_RATE);
        _pc2[_idx + 1] -= (_pc2[_idx + 1] >> SLOW_RATE);
        _pc2[_idx] -= (_pc2[_idx] >> SLOW_RATE);
    }
    else {
        _pc1[256] += ((_pc1[256] ^ 0xFFFF) >> FAST_RATE);
        _pc1[_c1] += ((_pc1[_c1] ^ 0xFFFF) >> MEDIUM_RATE);
        _pc2[_idx + 1] += ((_pc2[_idx + 1] ^ 0xFFFF) >> SLOW_RATE);
        _pc2[_idx] += ((_pc2[_idx] ^ 0xFFFF) >> SLOW_RATE);
        _ctx++;
    }

    if (_ctx > 255) {
        _c2 = _c1;
        _c1 = _ctx & 0xFF;
        _ctx = 1;

        if (_c1 == _c2) {
            _run++;
            _runMask = ((uint32)(2 - _run)) >> 31;
        }
        else {
            _run = 0;
            _runMask = 0;
        }
    }
}

// Return the split value representing the probability of 1 in the [0..4095] range.
inline int CMPredictor::get()
{
    _pc1 = _counter1[_ctx];
    const int p = (13 * _pc1[256] + 14 * _pc1[_c1] + 5 * _pc1[_c2]) >> 5;
    _idx = p >> 12;
    _pc2 = _counter2[(_ctx << 1) | _runMask];
    const int x1 = _pc2[_idx];
    const int x2 = _pc2[_idx + 1];
    const int ssep = x1 + (((x2 - x1) * (p & 4095)) >> 12);
    return (p + ssep + ssep + ssep + 32) >> 6; // rescale to [0..4095]
}

File: entropy/CMPredictor.go
/*
Copyright 2011-2021 Frederic Langlet
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
you may obtain a copy of the License at

                http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package entropy

const (
	_CM_FAST_RATE   = 2
	_CM_MEDIUM_RATE = 4
	_CM_SLOW_RATE   = 6
	_CM_PSCALE      = 65536
)

// CMPredictor context model predictor based on BCM by Ilya Muravyov.
// See https://github.com/encode84/bcm
type CMPredictor struct {
	c1       byte
	c2       byte
	ctx      int32
	idx      int
	runMask  int32
	counter1 [256][]int32
	counter2 [512][]int32
	p        int
}

// NewCMPredictor creates a new instance of CMPredictor
func NewCMPredictor() (*CMPredictor, error) {
	this := &CMPredictor{}
	this.ctx = 1
	this.runMask = 0

	for i := 0; i < 256; i++ {
		this.counter1[i] = make([]int32, 257)
		this.counter2[i+i] = make([]int32, 17)
		this.counter2[i+i+1] = make([]int32, 17)

		for j := 0; j <= 256; j++ {
			this.counter1[i][j] = _CM_PSCALE >> 1
		}

		for j := 0; j < 16; j++ {
			this.counter2[i+i][j] = int32(j << 12)
			this.counter2[i+i+1][j] = int32(j << 12)
		}

		this.counter2[i+i][16] = int32(15 << 12)
		this.counter2[i+i+1][16] = int32(15 << 12)
	}

	pc1 := this.counter1[this.ctx]
	this.p = int(13*(pc1[256]+pc1[this.c1])+6*pc1[this.c2]) >> 5
	this.idx = this.p >> 12
	return this, nil
}

// Update updates the probability model based on the internal bit counters
func (this *CMPredictor) Update(bit byte) {
	pc1 := this.counter1[this.ctx]
	pc2 := this.counter2[this.ctx|this.runMask]
	this.ctx += (this.ctx + int32(bit))

	if bit == 0 {
		pc1[256] -= (pc1[256] >> _CM_FAST_RATE)
		pc1[this.c1] -= (pc1[this.c1] >> _CM_MEDIUM_RATE)
		pc2[this.idx] -= (pc2[this.idx] >> _CM_SLOW_RATE)
		pc2[this.idx+1] -= (pc2[this.idx+1] >> _CM_SLOW_RATE)
	} else {
		pc1[256] -= ((pc1[256] - _CM_PSCALE + 16) >> _CM_FAST_RATE)
		pc1[this.c1] -= ((pc1[this.c1] - _CM_PSCALE + 16) >> _CM_MEDIUM_RATE)
		pc2[this.idx] -= ((pc2[this.idx] - _CM_PSCALE + 16) >> _CM_SLOW_RATE)
		pc2[this.idx+1] -= ((pc2[this.idx+1] - _CM_PSCALE + 16) >> _CM_SLOW_RATE)
	}

	if this.ctx > 255 {
		this.c2 = this.c1
		this.c1 = byte(this.ctx)
		this.ctx = 1

		if this.c1 == this.c2 {
			this.runMask = 0x100
		} else {
			this.runMask = 0
		}
	}

	pc1 = this.counter1[this.ctx]
	this.p = int(13*(pc1[256]+pc1[this.c1])+6*pc1[this.c2]) >> 5
	this.idx = this.p >> 12
}

// Get returns the value representing the probability of the next bit being 1
// in the [0..4095] range. The probability is computed from the internal
// bit counters.
func (this *CMPredictor) Get() int {
	pc2 := this.counter2[this.ctx|this.runMask]
	x2 := int(pc2[this.idx+1])
	x1 := int(pc2[this.idx])
	ssep := x1 + (((x2 - x1) * (this.p & 4095)) >> 12)
	return (this.p + 3*ssep + 32) >> 6 // rescale to [0..4095]
}
